engine/link/link.asm

```diff
@@ -2127,30 +2133,286 @@ Special_TryQuickSave:
 	ld [wChosenCableClubRoom], a
 	ret
 
-Special_CheckBothSelectedSameRoom:
-	ld a, [wChosenCableClubRoom]
-	call Link_EnsureSync
-	push af
-	call LinkDataReceived
+PerformLinkChecks:
+	xor a
+	ld bc, 10
+	ld hl, wLinkReceivedPolishedMiscBuffer
+	call ByteFill
+
+	; This acts as the old Special_CheckBothSelectedSameRoom
+	; We send a dummy byte here that will cause old versions
+	; of PolishedCrystal's CheckBothSelectedSameRoom function
+	; to fail.
+	ld a, LINK_ROOM_DUMMY - 1
+	call Link_ExchangeNybble
+	cp LINK_ROOM_DUMMY - 1
+	jmp nz, .OldVersionDetected
+
+	; Prepare for Multiple Byte Transfers
+	call CheckLinkTimeout_Gen2
+	ldh a, [hScriptVar]
+	and a
+	jmp z, LinkTimeout
+	ldh a, [hSerialConnectionStatus]
+	cp USING_INTERNAL_CLOCK
+	jr nz, .player_1
+
+	ld c, 3
+	call DelayFrames
+	xor a
+	ldh [hSerialSend], a
+	inc a
+	ldh [rSC], a
+	ld a, START_TRANSFER_INTERNAL_CLOCK
+	ldh [rSC], a
 	call DelayFrame
-	call LinkDataReceived
-	pop af
+	xor a
+	ldh [hSerialSend], a
+	inc a
+	ldh [rSC], a
+	ld a, START_TRANSFER_INTERNAL_CLOCK
+	ldh [rSC], a
+
+.player_1:
+	ld c, 3
+	call DelayFrames
+	ldh a, [rIF]
+	push af
+	ldh a, [rIE]
+	push af
+	xor a
+	ldh [rIF], a
+	ld a, 1 << SERIAL
+	ldh [rIE], a
```

Comment:
This is all mostly redundant code that also exists in `Gen2ToGen2LinkComms:`. I think i'm going to consolidate this part into a shared routine that both `Gen2ToGen2LinkComms:` and `PerformLinkChecks:` will both use. This should be the last thing to be done.